### ISSUE TITLE
Cannot connect to "Docker for Windows"

### ISSUE THREAD

**Issue by AndreLouisCaron** [terraform-providers/terraform-provider-docker#52]
Hi there,

I just tried using Terraform to provision some Docker containers for a local throwaway test infrastructure and I cannot manage to have Terraform connect to my Docker daemon.

### Terraform Version
```
Terraform v0.11.7
+ provider.docker v0.1.1
+ provider.template v1.0.0
```

### Affected Resource(s)
The provider's `host` attribute itself is affected.

### Terraform Configuration Files
```hcl
provider "docker" {}
resource "docker_container" "anything" { ... }
```

### Expected Behavior
I was hoping to create a simple Docker container :-)

### Actual Behavior
Terraform cannot connect to my Docker daemon.  I cannot find any suitable value for the `host` property that will allow it to connect.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Access a Windows 10 machine with "Docker for Windows" (the one based on Hyper-V).
2. Write a Terraform configuration that creates a single Docker container;
3. Run `terraform plan`

### Important Factoids
When running Docker for Windows, the Docker daemon exposes a Windows named pipe rather than a TCP endpoint.

**Comment by AndreLouisCaron** [comment]
**Note**: if I leave the `host` property empty, I get the following error:

```
Error: Error running plan: 1 error(s) occurred:

* provider.docker: Error pinging Docker server: Get http://unix.sock/_ping: dial unix /var/run/docker.sock: socket: An address incompatible with the requested protocol was used.
```

Seems like the default on Windows 10 is `http://unix.sock/`, which is unlikely to be correct :-)

**Comment by AndreLouisCaron** [comment]
**Note**: if I set the `host` property like the examples in the docs, I get the following error:

```
Error: Error running plan: 1 error(s) occurred:

* provider.docker: Error pinging Docker server: Get http://127.0.0.1:2376/_ping: dial tcp 127.0.0.1:2376: connectex: No connection could be made because the target machine actively refused it.
```

**Comment by AndreLouisCaron** [comment]
**Note**: if I set the `host` property to anything that looks like a named pipe endpoint (e.g. `\\.\pipe\docker_engine` or `npipe://...`, I get the following error:

```
Error: Error running plan: 1 error(s) occurred:

* provider.docker: Error initializing Docker client: invalid endpoint
```

**Comment by mavogel** [comment]
@AndreLouisCaron it's maybe something to do with the underlying library which initializes the client for windows https://github.com/fsouza/go-dockerclient/blob/master/client_windows.go#L29 . But now it seems like a good point to add `appveyor` for windows testing. I'll investigate

**Comment by AndreLouisCaron** [comment]
@mavogel Thanks for the quick reply!

I took a quick look at the code and it seems like the error is coming from `go-dockerclient`'s `parseEndpoint()` function.

We can see that the version that is vendored here ([ref](https://github.com/terraform-providers/terraform-provider-docker/blob/v0.1.1/vendor/github.com/fsouza/go-dockerclient/client.go#L905)) does not support Windows named pipe, whereas the same function in the latest version of `go-dockerclient` supports them ([ref](https://github.com/fsouza/go-dockerclient/blob/v1.2.0/client.go#L975)).

It seems to me like upgrading  `go-dockerclient` would fix my problem.

I'm more than willing to help out with a patch, but I'm not very familiar with golang and your vendoring process.  Any chance you can guide me through that?

Thanks!

**Comment by mavogel** [comment]
@AndreLouisCaron sure I can guide you through. Assuming you have `go` installed and `GO_PATH` set

Update the dependency by running the following with the desired `ref` command in the `terraform-provider-docker` folder
```sh
go get -u github.com/kardianos/govendor
govendor fetch github.com/fsouza/go-dockerclient@<ref>
# if this works on your windows machine. It assures the deps are all fine
make vendor-status
make vet
```

Is is possible in the meantime to run `make` on windows? If not, it would be lovely of you could support a `bat` file which set everything up or even better a configuration for `appveyor`  😃 Unfortunately I do not have any windows machine here handy to do it on my own...

**Comment by AndreLouisCaron** [comment]
@mavogel Thanks for the info!

I'm not sure I will do the "right thing" on Windows.  Assuming that the library update is desirable (regardless of this specific issue), I can try doing this update on my mac machine and then I can test it again on Windows after the update.

Is that fine with you?

**Comment by AndreLouisCaron** [comment]
OK, I put up a PR with the library upgrade.  I'm trying to test this against Docker for Windows and I'm having a bit of trouble.

I'm using this terraform config:

```
provider "docker" {
  host = "npipe://./pipe/docker_engine"
}
resource "docker_container" "redis" {
  image = "redis:3.2"
  name = "test_redis"
}
```

I ran `terraform init` successfully, but `terraform plan` still fails with that "invalid endpoint" error.

I followed the instructions in the README: `go build`, `terraform init`, etc.

I even added a new test in `provider_test.go`:

```go

func TestProviderNamedPipe(t *testing.T) {
	config := DockerConfig{
		Host:     "npipe://./pipe/docker_engine",
		Ca:       "",
		Cert:     "",
		Key:      "",
		CertPath: "",
	}
	client, err := config.NewClient()
	if err != nil {
		t.Fatalf("err: %s", err)
	}
	client.Ping()
}
```

This test passes successfully, suggesting that the new version of the libary works fine.

I'm not sure if terraform is picking up my custom build of the provider.  I double-checked that `%GOPATH%\bin` contains `terraform-provider-docker` and that I re-ran `terraform init` after building.

I feel like I'm not far off.  Any ideas?

**Comment by mavogel** [comment]
The test passes because it is not an acceptance test. Rename to 
```go
func TestAccProviderNamedPipe(t *testing.T) {
```
and try again please but with the env `TF_ACC=1` set. Otherwise the acceptance test will not be run.
I think also you'd have to set to make the acceptance tests to run
```
export DOCKER_REGISTRY_ADDRESS="127.0.0.1:15000"
export DOCKER_REGISTRY_USER="testuser"
export DOCKER_REGISTRY_PASS="testpwd"
export DOCKER_PRIVATE_IMAGE="127.0.0.1:15000/tftest-service:v1"
```

Additionally you can also run `make testacc`. But it's linux based. If you could provide a `bat` file for `runAccTests.sh`, I'd be pleased :)

**Comment by mavogel** [comment]
@AndreLouisCaron just FYI: I'm currently migrating to the official docker client because I cannot build the provider. See https://github.com/fsouza/go-dockerclient/issues/733 for details.

**Comment by mavogel** [comment]
@AndreLouisCaron I just released `1.0.0` with the offical docker client as api client. Can you try again if this issue still occurs?

**Comment by mavogel** [comment]
Added functionality and fixes in #90
